### ISSUE TITLE
[red-knot] support named tuples and typing.NamedTuple

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#16994]

**Comment by bswck** [comment]
`super()` is unsupported in `typing.NamedTuple`s--would be cool to also catch that (https://github.com/python/cpython/pull/130082)

**Comment by dcreager** [comment]
https://github.com/astral-sh/ruff/pull/16538 currently adds a bunch of new `NamedTuple`-related false positives.

`NamedTuple` can be used either as a base class or as a function:

```py
class A(NamedTuple):
    x: int
    y: str

A = NamedTuple("A", [("x", int), ("y", str)])
```

typeshed supports the latter by adding some `__init__` overloads to `NamedTuple`:

```py
@overload
def __init__(self, typename: str, fields: Iterable[tuple[str, Any]], /) -> None: ...

@overload
def __init__(self, typename: str, fields: None = None, /, **kwargs: Any) -> None: ...
```

For the former, we'll probably need to reuse some of the `dataclass` machinery to create a custom `__init__` method for each `NamedTuple` subclass.

**Comment by carljm** [comment]
We can also temporarily eliminate all these false positives with a simple special-case handling for instantiating a named tuple that accepts any arguments; we had a similar special-case in place for dataclass before we handled its constructor properly. (Though probably doing this correctly for NamedTuple by emulating our existing dataclass support wouldn't be too hard?)

**Comment by carljm** [comment]
Marking this for alpha simply due to the quantity of false positives it causes in the ecosystem; the alpha "fix" can be simply the special case to silence all the false positives for now.

**Comment by bswck** [comment]
I know this isn't forced by standard, but I'm also advocating for [namedtuples to be final](https://github.com/bswck/make-typed-namedtuples-final). You may or may not want to check that. Or disrecommend that with a linter, if that makes sense.

**Comment by sharkdp** [comment]
Initial support for `NamedTuple` was merged in #17738. I'm closing this for now. There are some open TODOs in the test suite and some ideas in the issue here, but I'm unsure if it's worth opening a new ticket just yet. I think we'll automatically be forced to implement more functionality eventually, either by bug reports, ecosystem false positives or when we add support for running the typing conformance checks.
